#include "mainwindow.h"
#include "../backend_headers/FileManager.h"
#include "ui_mainwindow.h"
#include <QCloseEvent>
#include <QIcon>
#include <QListWidgetItem>
#include <QMessageBox>

#include "adminpanelwidget.h"
#include "dashboardwidget.h"
#include "productmanagementwidget.h"
#include "receiptmanagementwidget.h"
#include "reportwidget.h"
#include "suppliermanagementwidget.h"

MainWindow::MainWindow(const NguoiDung *user, QWidget *parent)
    : QMainWindow(parent), m_closeAction(Shutdown), ui(new Ui::MainWindow), m_currentUser(user) {
    ui->setupUi(this);

    loadAllData();
    setupNavigation();
    setupContentWidgets();

    connect(ui->navWidget, &QListWidget::currentRowChanged, this,
            &MainWindow::on_navWidget_currentRowChanged);

    if (m_currentUser) {
        ui->welcomeLabel->setText(
            QString("Xin chÃ o, %1")
                .arg(QString::fromStdString(m_currentUser->getHoTen())));
    }

    ui->navWidget->setCurrentRow(0);
}

MainWindow::~MainWindow() { delete ui; }

MainWindow::CloseAction MainWindow::getCloseAction() const {
    return m_closeAction;
}
void MainWindow::closeEvent(QCloseEvent *event) {
    /*if (m_closeAction == Logout) {
        saveAllData();
        event->accept();
        return;
    }*/
    QMessageBox::StandardButton resBtn = QMessageBox::question(
        this, "XÃ¡c nháº­n thoÃ¡t",
        "Báº¡n cÃ³ muá»‘n lÆ°u láº¡i táº¥t cáº£ thay Ä‘á»•i trÆ°á»›c khi thoÃ¡t?",
        QMessageBox::Cancel | QMessageBox::No | QMessageBox::Yes,
        QMessageBox::Yes);
    if (resBtn == QMessageBox::Yes) {
        saveAllData();
        event->accept();
        qApp->quit();
    } else if (resBtn == QMessageBox::No) {
        event->accept();
        qApp->quit();
    } else {
        event->ignore();
    }
}

void MainWindow::loadAllData() {
    FileManager fm;

    fm.docTatCaDuLieu(m_qlhh, m_qlncc, m_qlp, m_qlnd);
    ui->statusbar->showMessage("Táº£i dá»¯ liá»‡u thÃ nh cÃ´ng.", 3000);
}

void MainWindow::saveAllData() {
    FileManager fm;
    fm.ghiTatCaDuLieu(m_qlhh, m_qlncc, m_qlp, m_qlnd);
    ui->statusbar->showMessage("Äang lÆ°u dá»¯ liá»‡u...", 2000);
    ui->statusbar->showMessage("LÆ°u dá»¯ liá»‡u thÃ nh cÃ´ng.", 3000);
}

void MainWindow::setupNavigation() {
    ui->navWidget->addItem(
        new QListWidgetItem(QIcon(":/icons/dashboard.png"), "ðŸ“ˆ Tá»•ng Quan"));
    ui->navWidget->addItem(new QListWidgetItem(QIcon(":/icons/product.png"),
                                               "ðŸ“¦ Quáº£n lÃ½ HÃ ng HÃ³a"));
    ui->navWidget->addItem(
        new QListWidgetItem(QIcon(":/icons/supplier.png"), "ðŸšš Quáº£n lÃ½ NCC"));
    ui->navWidget->addItem(
        new QListWidgetItem(QIcon(":/icons/import.png"), "ðŸ“¥ Quáº£n lÃ½ Phiáº¿u"));
    ui->navWidget->addItem(
        new QListWidgetItem(QIcon(":/icons/report.png"), "ðŸ“Š BÃ¡o cÃ¡o & TK"));

    if (m_currentUser && m_currentUser->getVaiTro() == "quanly") {
        ui->navWidget->addItem(new QListWidgetItem(QIcon(":/icons/admin.png"),
                                                   "âš™ï¸ Quáº£n trá»‹ Há»‡ thá»‘ng"));
    }
}

void MainWindow::setupContentWidgets() {
    ui->contentStackedWidget->addWidget(
        new DashboardWidget(m_qlhh, m_qlp, this));
    ui->contentStackedWidget->addWidget(
        new ProductManagementWidget(m_qlhh, m_qlncc, this));
    ui->contentStackedWidget->addWidget(
        new SupplierManagementWidget(m_qlncc, this));
    ui->contentStackedWidget->addWidget(
        new ReceiptManagementWidget(m_qlp, m_qlhh, m_qlncc, this));
    ui->contentStackedWidget->addWidget(new ReportWidget(m_qlp, m_qlhh, this));

    if (m_currentUser && m_currentUser->getVaiTro() == "quanly") {
        ui->contentStackedWidget->addWidget(
            new AdminPanelWidget(m_qlnd, m_currentUser, this));
    }
}

void MainWindow::on_navWidget_currentRowChanged(int currentRow) {
    if (currentRow >= 0 && currentRow < ui->contentStackedWidget->count()) {
        ui->contentStackedWidget->setCurrentIndex(currentRow);
    }
}

void MainWindow::on_logoutButton_clicked() {
    m_closeAction = Logout;
    this->close();
}
